### PR TITLE
[WIP] traits/fulfill: track `GenericArgs`s instead of `InferTy`s in `stalled_on`.

### PR DESCRIPTION
**Pull request by eddyb**
**EDIT**: superseded by #70213 for performance reasons.

This PR addresses the representation side of #70180, but only *actually collects* `ty::Infer`s via `Ty::walk` into `stalled_on` (collecting `ty::ConstKind::Infer`s requires #70164).

However, it should be enough to handle #70107's needs (WF obligations are stalled only on the outermost type/const being an inference variable, no `walk`-ing is involved).

*This PR will make step-by-step changes in order to check the perf impact of each.*

r? @nikomatsakis cc @nnethercote

### CODE REVIEW

**Comment by eddyb** [comment]
I'm doing a perf run with just this first commit, to assess the impact of having to match on `Ty` to get at the `ty::InferTy` inside each `stalled_on` entry:
> traits/fulfill: track full `Ty`s instead of `InferTy`s in `stalled_on`.

@bors try @rust-timer queue

**Comment by Centril** [comment]
Perf seems mostly *improved* I'd say, but it's a bit of a wash.

**Comment by eddyb** [comment]
@Centril most of the perf improvements around 1% I've seen lately look dubious and unrelated (i.e. likely noise), so I focus on the regressions and try to find a relevant one.

Besides, it should be impossible for this PR to make anything faster, so there's that.

[`keccak-opt` clean](https://perf.rust-lang.org/detailed-query.html?commit=fc63e0e7ac01afe8ebf2cbb38466eb80c1d42922&base_commit=f4c675c476c18b1a11041193f2f59d695b126bc8&benchmark=keccak-opt&run_name=clean) and [`inflate-check` clean](https://perf.rust-lang.org/detailed-query.html?commit=fc63e0e7ac01afe8ebf2cbb38466eb80c1d42922&base_commit=f4c675c476c18b1a11041193f2f59d695b126bc8&benchmark=inflate-check&run_name=clean) paint a pretty clear picture: type-checking is slower.

I think the efficient approach moving forward would be to make an `enum StalledOnInferVar` with these variants (`Ty`, `Int` and `Float`) plus a `Const` one: https://github.com/rust-lang/rust/blob/1057dc97afce39ff6a224966ece3ed438af4c1f5/src/librustc_infer/infer/mod.rs#L1631-L1655

I'm closing this PR to serve as a tombstone for the inefficient approach.

File: src/librustc_trait_selection/traits/fulfill.rs [hunk 1]
```diff
@@ -534,20 +543,19 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
...
 
-/// Returns the set of type variables contained in a trait ref
-fn trait_ref_type_vars<'a, 'tcx>(
+/// Returns the set of inference variables contained in a trait ref.
+fn trait_ref_infer_vars<'a, 'tcx>(
     selcx: &mut SelectionContext<'a, 'tcx>,
-    t: ty::PolyTraitRef<'tcx>,
-) -> Vec<ty::InferTy> {
-    t.skip_binder() // ok b/c this check doesn't care about regions
+    ty: ty::PolyTraitRef<'tcx>,
```
Comment (eddyb) [inline src/librustc_trait_selection/traits/fulfill.rs]: Oops, I should've renamed this to `trait_ref`.

**Pull request closed without merging**